packages/@aws-cdk/core/lib/deployment-environment.ts

```diff
@@ -0,0 +1,532 @@
+import * as asset_schema from '@aws-cdk/cdk-assets-schema';
+import * as cxapi from '@aws-cdk/cx-api';
+import * as fs from 'fs';
+import * as path from 'path';
+import { DockerImageAssetLocation, DockerImageAssetSource, FileAssetLocation, FileAssetSource } from "./assets";
+import { Fn } from './cfn-fn';
+import { Construct, ISynthesisSession } from "./construct";
+import { FileAssetParameters } from './private/asset-parameters';
+import { TemplatedString } from './private/templated-string';
+import { Stack } from "./stack";
+import { Token } from './token';
+
+/**
+ * The well-known name for the docker image asset ECR repository. All docker
+ * image assets will be pushed into this repository with an image tag based on
+ * the source hash.
+ */
+const ASSETS_ECR_REPOSITORY_NAME = "aws-cdk/assets";
+
+/**
+ * This allows users to work around the fact that the ECR repository is
+ * (currently) not configurable by setting this context key to their desired
+ * repository name. The CLI will auto-create this ECR repository if it's not
+ * already created.
+ */
+const ASSETS_ECR_REPOSITORY_NAME_OVERRIDE_CONTEXT_KEY = "assets-ecr-repository-name";
+
+/**
+ * Encodes information how a certain Stack should be deployed
+ */
+export interface IDeploymentEnvironment {
+  /**
+   * Bind to the stack this environment is going to be used on
+   *
+   * Must be called before any of the other methods are called.
+   */
+  bind(stack: Stack): void;
+
+  /**
+   * Register a File Asset
+   *
+   * Returns the parameters that can be used to refer to the asset inside the template.
+   */
+  addFileAsset(asset: FileAssetSource): FileAssetLocation;
+
+  /**
+   * Register a Docker Image Asset
+   *
+   * Returns the parameters that can be used to refer to the asset inside the template.
+   */
+  addDockerImageAsset(asset: DockerImageAssetSource): DockerImageAssetLocation;
+
+  /**
+   * Access the stack deployment configuration
+   */
+  stackDeploymentConfig(): StackDeploymentConfig;
+
+  /**
+   * Synthesize additional artifacts into the session
+   *
+   * @experimental
+   */
+  synthesize(session: ISynthesisSession): DeploymentEnvironmentSynthesisResult;
```

Comment:
Maybe `synthesize()` is the wrong name to use, but there is definitely an interplay between this and the stack where this may add artifacts to the deployment session that the stack does not have to know about.

See asset manifest.